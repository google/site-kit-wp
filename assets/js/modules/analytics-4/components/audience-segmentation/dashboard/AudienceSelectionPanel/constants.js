/**
 * AudienceSelectionPanel components - constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AUDIENCE_ADD_GROUP_NOTICE_SLUG =
	'audience-segmentation-add-group-notice';
export const AUDIENCE_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-audience-selection-panel-opened';
export const AUDIENCE_SELECTION_FORM = 'audience-selection-form';
export const AUDIENCE_SELECTED = 'audience-selected';
export const AUDIENCE_SELECTION_CHANGED = 'audience-selection-changed';
export const AUDIENCE_CREATION_NOTICE_SLUG =
	'audience-segmentation-creation-notice';
export const AUDIENCE_CREATION_SUCCESS_NOTICE_SLUG =
	'audience-segmentation-creation-success-notice';
export const MIN_SELECTED_AUDIENCES_COUNT = 1;
export const MAX_SELECTED_AUDIENCES_COUNT = 3;
export const AUDIENCE_CREATION_EDIT_SCOPE_NOTICE_SLUG =
	'audience-creation-edit-scope-notice';
export const AUDIENCE_CREATION_FORM = 'audience-creation-form';
