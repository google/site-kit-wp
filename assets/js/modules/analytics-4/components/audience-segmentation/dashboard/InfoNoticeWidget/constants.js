/**
 * Constants related to audience segmentation info notices.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * UI key to display audience segmentation info notice.
 *
 * @since 1.135.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_HIDE_UI =
	'audience-segmentation-info-notice-ui';

/**
 * Slug for audience segmentation info notice.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICE_SLUG = 'audience-segmentation-info-notice';

// Notice slugs: new-visitors, compare-metrics, custom-audiences, purchasers, returning-visitors, compare-new-returning, or compare-cities

/**
 * List of info notices for audiences.
 *
 * @since 1.130.0
 * @private
 */
export const AUDIENCE_INFO_NOTICES = [
	{
		slug: 'new-visitors',
		content: __(
			'The higher the portion of new visitors you have, the more your audience is growing. Looking at what content brings them to your site may give you insights on how to reach even more people.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-metrics',
		content: __(
			'Select up to three visitor groups to display on the dashboard and easily compare metrics between them.',
			'google-site-kit'
		),
	},
	{
		slug: 'custom-audiences',
		content: __(
			'Configure your own custom audiences in Analytics to gain deeper insights into visitor behavior, for example consider creating an “Existing customers” or “Subscribers” segment, depending on what goals you have for your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'purchasers',
		content: __(
			'Select the Purchasers visitor group to gain insights into which visitors bring the most revenue to your site.',
			'google-site-kit'
		),
	},
	{
		slug: 'returning-visitors',
		content: __(
			'The more returning visitors your site has, the stronger and more loyal an audience you’re building. Check which content brings people back to your site - it might help you create a strategy to build a community.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-new-returning',
		content: __(
			'Compare the ratio of “new” to “returning” visitors – this can give you insights on whether you have more people stopping by as a one-off, or more loyal visitors.',
			'google-site-kit'
		),
	},
	{
		slug: 'compare-cities',
		content: __(
			'Check the cities which bring you more new vs more returning visitors – there might be new audiences you could engage with in locations you hadn’t thought about.',
			'google-site-kit'
		),
	},
];
