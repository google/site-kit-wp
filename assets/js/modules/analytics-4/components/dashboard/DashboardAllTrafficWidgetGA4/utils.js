/**
 * DashboardAllTrafficWidgetGA4 utility functions.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { renderToString } from '@wordpress/element';

/**
 * Internal dependencies
 */
import Link from '../../../../../components/Link';
import Typography from '../../../../../components/Typography';

/**
 * Creates a GA4 report row for the given date, with a zero for the metric value.
 *
 * @since 1.96.0
 * @since 1.123.0 Migrated from analytics to analytics-4.
 *
 * @param {string} date Date in YYYY-MM-DD format.
 * @return {Object} Row object with the date in the GA4 report format as the dimension value, and a zero for the metric value.
 */
export function createZeroDataRow( date ) {
	// Note, this uses `replace()` rather than `replaceAll()` as our current version of Node does not support `replaceAll()`.
	// TODO: Replace with `replaceAll()` when we upgrade to Node 15+.
	return {
		dimensionValues: [
			{
				value: date.replace( new RegExp( '-', 'g' ), '' ),
			},
		],
		metricValues: [
			{
				value: 0,
			},
		],
	};
}
/**
 * Gets HTML string for a tooltip help link.
 *
 * @since 1.116.0
 * @since 1.123.0 Migrated from analytics to analytics-4.
 *
 * @param {string} url      The URL to link to.
 * @param {string} label    The label for the link.
 * @param {string} rowLabel The label for the row.
 * @return {string} The HTML string for the tooltip help link.
 */
export const getTooltipHelp = ( url, label, rowLabel ) =>
	renderToString(
		<Typography as="p" size="medium" type="body">
			<Link
				className="googlesitekit-cta-link__tooltip"
				href={ url }
				data-row-label={ rowLabel }
				external
			>
				{ label }
			</Link>
		</Typography>
	);
