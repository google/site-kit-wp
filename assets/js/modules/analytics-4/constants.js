/**
 * GA4 module constants.
 *
 * Site Kit by Google, Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const ACTIVATION_STEP_REMINDER = 0;
export const ACTIVATION_STEP_SETUP = 1;
export const ACTIVATION_STEP_SUCCESS = 2;
export const ACTIVATION_STEP_IN_PROGRESS = 3;

export const ACTIVATION_ACKNOWLEDGEMENT_TOOLTIP_STATE_KEY =
	'activation-acknowledgement-tooltip-state';

export const GA4_ACTIVATION_BANNER_STATE_KEY = 'ga4-activation-banner-state';

export const ENHANCED_MEASUREMENT_ACTIVATION_BANNER_TOOLTIP_STATE_KEY =
	'enhanced-measurement-activation-banner-tooltip-state';
export const LEGACY_ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY =
	'enhanced-measurement-activation-banner-dismissed-item';

export const REPORT_ARGS_SELECTED_METRIC_KEY = '_r.explorerCard..selmet';
export const REPORT_ARGS_SELECTED_DIMENSION_KEY = '_r.explorerCard..seldim';
export const REPORT_ARGS_DATA_FILTERS_KEY = '_r..dataFilters';
export const REPORT_ARGS_NAV_KEY = '_r..nav';

export const KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY =
	'key-metrics-connect-ga4-cta-widget';

/**
 * Date that Site Kit will automatically switch to using GA4: September 25, 2023.
 *
 * With Site Kit's maximum date range being three months, once this period has passed since the UA cutoff date,
 * it will no longer be able to show any historical data for UA and will automatically switch to GA4.
 * For operational reasons, we automatically make the switch a few days before the three month cutoff point.
 *
 * @since 1.107.0
 * @see UA_CUTOFF_DATE.
 */
export const GA4_AUTO_SWITCH_DATE = '2023-09-25';
export const MODULE_SLUG_ANALYTICS_4 = 'analytics-4';
