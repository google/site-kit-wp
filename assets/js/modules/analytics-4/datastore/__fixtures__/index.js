/**
 * GA4 Datastore Fixtures.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as accountSummaries } from './account-summaries.json';
export { default as audiences } from './audiences.json';
export { default as availableAudiences } from './available-audiences.json';
export { default as container } from './container.json';
export { default as containerE2E } from './container-e2e.json';
export { default as containerDestinations } from './container-destinations.json';
export { default as conversionEvents } from './conversion-events.json';
export { default as createProperty } from './create-property.json';
export { default as createWebDataStream } from './create-webdatastream.json';
export { default as defaultSettings } from './default-settings.json';
export { default as defaultEnhancedMeasurementSettings } from './default-enhanced-measurement-settings.json';
export { default as pageTitles } from './page-titles.json';
export { default as properties } from './properties.json';
export { default as report } from './report.json';
export { default as pivotReport } from './pivot-report.json';
export { default as webDataStreams } from './webdatastreams.json';
export { default as webDataStreamsBatch } from './webdatastreams-batch.json';
export { default as webDataStreamsBatchSetup } from './webdatastreams-batch-setup.json';
export { default as googleTagSettings } from './google-tag-settings.json';
