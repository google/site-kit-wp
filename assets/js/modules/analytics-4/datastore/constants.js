/**
 * `modules/analytics-4` data store constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const MODULES_ANALYTICS_4 = 'modules/analytics-4';

// A special Account ID value used for the "Set up a new account" option.
export const ACCOUNT_CREATE = 'account_create';

// A special Property ID value used for the "Set up a new property" option.
export const PROPERTY_CREATE = 'property_create';

export const WEBDATASTREAM_CREATE = 'webdatastream_create';

export const FORM_SETUP = 'analyticsSetup';

export const MAX_WEBDATASTREAMS_PER_BATCH = 10;

// Date range offset days for Analytics 4 report requests.
export const DATE_RANGE_OFFSET = 1;

export const GTM_SCOPE = 'https://www.googleapis.com/auth/tagmanager.readonly';

export const ENHANCED_MEASUREMENT_FORM = 'enhanced-measurement-form';
export const ENHANCED_MEASUREMENT_ENABLED = 'enhanced-measurement-enabled';
export const ENHANCED_MEASUREMENT_SHOULD_DISMISS_ACTIVATION_BANNER =
	'enhanced-measurement-should-dismiss-activation-banner';

// Form ID for the account creation form.
export const FORM_ACCOUNT_CREATE = 'analyticsAccountCreate';

export const FORM_CUSTOM_DIMENSIONS_CREATE = 'analyticsCustomDimensionsCreate';

// OAuth scope required for provisioning a Google Analytics account.
export const PROVISIONING_SCOPE =
	'https://www.googleapis.com/auth/analytics.provision';
export const EDIT_SCOPE = 'https://www.googleapis.com/auth/analytics.edit';

// Dashboard widget constants.
export const UI_DIMENSION_NAME = 'dashboardAllTrafficWidgetDimensionName';
export const UI_DIMENSION_COLOR = 'dashboardAllTrafficWidgetDimensionColor';
export const UI_DIMENSION_VALUE = 'dashboardAllTrafficWidgetDimensionValue';
export const UI_ACTIVE_ROW_INDEX = 'dashboardAllTrafficWidgetActiveRowIndex';
export const UI_ALL_TRAFFIC_LOADED = 'dashboardAllTrafficWidgetLoaded';

// Note: names and descriptions are not translated as these are not surfaced in Site Kit
// and are also subject to hard limits on the length which would be unpredictable if translated.
// See https://developers.google.com/analytics/devguides/config/admin/v1/rest/v1beta/properties.customDimensions#CustomDimension
export const CUSTOM_DIMENSION_DEFINITIONS = {
	googlesitekit_post_date: {
		parameterName: 'googlesitekit_post_date',
		displayName: 'WordPress Post Date',
		description: 'Created by Site Kit: Date when a post was published',
		scope: 'EVENT',
	},
	googlesitekit_post_author: {
		parameterName: 'googlesitekit_post_author',
		displayName: 'WordPress Post Author',
		description: 'Created by Site Kit: WordPress name of the post author',
		scope: 'EVENT',
	},
	googlesitekit_post_categories: {
		parameterName: 'googlesitekit_post_categories',
		displayName: 'WordPress Post Categories',
		description:
			'Created by Site Kit: Names of categories assigned to a post',
		scope: 'EVENT',
	},
	googlesitekit_post_type: {
		parameterName: 'googlesitekit_post_type',
		displayName: 'WordPress Post Type',
		description: 'Created by Site Kit: Content type of a post',
		scope: 'EVENT',
	},
};

export const ENUM_CONVERSION_EVENTS = {
	ADD_TO_CART: 'add_to_cart',
	PURCHASE: 'purchase',
	SUBMIT_LEAD_FORM: 'submit_lead_form',
	GENERATE_LEAD: 'generate_lead',
	CONTACT: 'contact',
};
export const CONVERSION_REPORTING_LEAD_EVENTS = [
	ENUM_CONVERSION_EVENTS.CONTACT,
	ENUM_CONVERSION_EVENTS.GENERATE_LEAD,
	ENUM_CONVERSION_EVENTS.SUBMIT_LEAD_FORM,
];

// Audience enums.
export const AUDIENCE_FILTER_CLAUSE_TYPE_ENUM = {
	AUDIENCE_CLAUSE_TYPE_UNSPECIFIED: 'AUDIENCE_CLAUSE_TYPE_UNSPECIFIED',
	INCLUDE: 'INCLUDE',
	EXCLUDE: 'EXCLUDE',
};

export const AUDIENCE_FILTER_SCOPE_ENUM = {
	AUDIENCE_FILTER_SCOPE_UNSPECIFIED: 'AUDIENCE_FILTER_SCOPE_UNSPECIFIED',
	AUDIENCE_FILTER_SCOPE_WITHIN_SAME_EVENT:
		'AUDIENCE_FILTER_SCOPE_WITHIN_SAME_EVENT',
	AUDIENCE_FILTER_SCOPE_WITHIN_SAME_SESSION:
		'AUDIENCE_FILTER_SCOPE_WITHIN_SAME_SESSION',
	AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS:
		'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
};

export const SITE_KIT_AUDIENCE_DEFINITIONS = {
	'new-visitors': {
		description: 'People who visited the site for the first time',
		displayName: 'New visitors',
		membershipDurationDays: -1, // The longest duration, 540 days.
		filterClauses: [
			{
				clauseType: 'INCLUDE',
				simpleFilter: {
					scope: 'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
					filterExpression: {
						andGroup: {
							filterExpressions: [
								{
									orGroup: {
										filterExpressions: [
											{
												dimensionOrMetricFilter: {
													fieldName: 'newVsReturning',
													stringFilter: {
														matchType: 'EXACT',
														value: 'new',
													},
												},
											},
										],
									},
								},
								{
									orGroup: {
										filterExpressions: [
											{
												notExpression: {
													dimensionOrMetricFilter: {
														fieldName: 'groupId',
														stringFilter: {
															matchType: 'EXACT',
															value: 'created_by_googlesitekit:new_visitors',
														},
													},
												},
											},
										],
									},
								},
							],
						},
					},
				},
			},
		],
	},
	'returning-visitors': {
		description: 'People who have visited your site at least once before',
		displayName: 'Returning visitors',
		membershipDurationDays: -1, // The longest duration, 540 days.
		filterClauses: [
			{
				clauseType: 'INCLUDE',
				simpleFilter: {
					scope: 'AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS',
					filterExpression: {
						andGroup: {
							filterExpressions: [
								{
									orGroup: {
										filterExpressions: [
											{
												dimensionOrMetricFilter: {
													fieldName: 'newVsReturning',
													stringFilter: {
														matchType: 'EXACT',
														value: 'returning',
													},
												},
											},
										],
									},
								},
								{
									orGroup: {
										filterExpressions: [
											{
												notExpression: {
													dimensionOrMetricFilter: {
														fieldName: 'groupId',
														stringFilter: {
															matchType: 'EXACT',
															value: 'created_by_googlesitekit:returning_visitors',
														},
													},
												},
											},
										],
									},
								},
							],
						},
					},
				},
			},
		],
	},
};

export const AUDIENCE_SEGMENTATION_SETUP_FORM = 'audiencePermissionsSetup';

export const AUDIENCE_TILE_CUSTOM_DIMENSION_CREATE =
	'audienceTileCustomDimensionCreate';

export const AUDIENCE_TYPE_SORT_ORDER = {
	USER_AUDIENCE: 0,
	SITE_KIT_AUDIENCE: 1,
	DEFAULT_AUDIENCE: 2,
};

export const AUDIENCE_ITEM_NEW_BADGE_SLUG_PREFIX =
	'audience-selection-panel-expirable-new-badge-';
