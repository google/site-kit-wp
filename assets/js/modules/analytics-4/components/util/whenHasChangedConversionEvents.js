/**
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';

/**
 * Higher-Order Component to render wrapped components when changed conversion events are detected.
 *
 * @since n.e.x.t
 *
 * @return {Function} Enhancing function.
 */
export default function whenHasChangedConversionEvents() {
	return ( WrappedComponent ) => {
		function WhenHasChangedConversionEvents( props ) {
			const hasChangedConversionEvents = useSelect( ( select ) => {
				const {
					hasNewConversionReportingEvents,
					hasLostConversionReportingEvents,
				} = select( MODULES_ANALYTICS_4 );

				return (
					hasNewConversionReportingEvents() ||
					hasLostConversionReportingEvents()
				);
			}, [] );

			if ( hasChangedConversionEvents ) {
				return <WrappedComponent { ...props } />;
			}

			const { WidgetNull = null } = props;
			return WidgetNull && <WidgetNull />;
		}

		return WhenHasChangedConversionEvents;
	};
}
