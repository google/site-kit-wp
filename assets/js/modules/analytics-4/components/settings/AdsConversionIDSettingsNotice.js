/**
 * AdsConversionIDSettingsNotice component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { createInterpolateElement, useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import { ADS_CONVERSION_ID_NOTICE_DISMISSED_ITEM_KEY } from '../../constants';
import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { WEEK_IN_SECONDS, dateSub, trackEvent } from '../../../../util';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';
import SettingsNotice, {
	TYPE_INFO,
} from '../../../../components/SettingsNotice';
import InfoCircleIcon from '../../../../../../assets/svg/icons/info-circle.svg';
import Link from '../../../../components/Link';
import useViewContext from '../../../../hooks/useViewContext';
import withIntersectionObserver from '../../../../util/withIntersectionObserver';

const SettingsNoticeWithIntersectionObserver =
	withIntersectionObserver( SettingsNotice );

export default function AdsConversionIDSettingsNotice() {
	const settingsAdminURL = useSelect( ( select ) =>
		select( CORE_SITE ).getAdminURL( 'googlesitekit-settings' )
	);

	const shouldShowNotice = useSelect( ( select ) => {
		const isDismissed = select( CORE_USER ).isItemDismissed(
			ADS_CONVERSION_ID_NOTICE_DISMISSED_ITEM_KEY
		);
		if ( isDismissed ) {
			return false;
		}

		const adsConversionIDMigratedAtMs =
			select( MODULES_ANALYTICS_4 ).getAdsConversionIDMigratedAtMs();
		if ( ! adsConversionIDMigratedAtMs ) {
			return false;
		}

		// If it has been <= 28 days since the migration.
		const referenceDate = select( CORE_USER ).getReferenceDate();
		const fourWeeksAgo = dateSub( referenceDate, 4 * WEEK_IN_SECONDS );
		return fourWeeksAgo.getTime() <= adsConversionIDMigratedAtMs;
	} );

	const viewContext = useViewContext();
	const trackDismissNotificationEvent = () => {
		trackEvent(
			`${ viewContext }_GA_Ads_redirect`,
			'dismiss_notification'
		);
	};
	const trackConfirmNotificationEvent = () => {
		trackEvent(
			`${ viewContext }_GA_Ads_redirect`,
			'confirm_notification'
		);
	};

	const onInView = useCallback( () => {
		trackEvent( `${ viewContext }_GA_Ads_redirect`, 'view_notification' );
	}, [ viewContext ] );

	// Do not show the notice if the view conditions have not been met.
	if ( ! shouldShowNotice ) {
		return null;
	}

	return (
		<SettingsNoticeWithIntersectionObserver
			onInView={ onInView }
			className="googlesitekit-settings-analytics-ads-conversion-id-notice"
			dismiss={ ADS_CONVERSION_ID_NOTICE_DISMISSED_ITEM_KEY }
			dismissCallback={ trackDismissNotificationEvent }
			dismissLabel={ __( 'Got it', 'google-site-kit' ) }
			type={ TYPE_INFO }
			Icon={ InfoCircleIcon }
			notice={ createInterpolateElement(
				__(
					'Ads Conversion ID has been moved to <a>Ads settings</a>',
					'google-site-kit'
				),
				{
					a: (
						<Link
							href={ `${ settingsAdminURL }#/connected-services/ads` }
							onClick={ trackConfirmNotificationEvent }
						/>
					),
				}
			) }
		/>
	);
}
