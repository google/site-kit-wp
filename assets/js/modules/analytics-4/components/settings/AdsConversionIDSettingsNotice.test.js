/**
 * AdsConversionIDSettingsNotice tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ADS_CONVERSION_ID_NOTICE_DISMISSED_ITEM_KEY } from '../../constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { DAY_IN_SECONDS, dateSub } from '../../../../util';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';
import AdsConversionIDSettingsNotice from './AdsConversionIDSettingsNotice';
import { render } from '../../../../../../tests/js/test-utils';
import { createTestRegistry } from '../../../../../../tests/js/utils';

describe( 'AdsConversionIDSettingsNotice', () => {
	let registry;
	let referenceDate;

	beforeEach( () => {
		registry = createTestRegistry();
		registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );

		referenceDate = registry.select( CORE_USER ).getReferenceDate();
	} );

	it( 'should not render if the migration has not been performed', () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );

		const { container } = render( <AdsConversionIDSettingsNotice />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should not render if it has been over 28 days since the migration', () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			adsConversionIDMigratedAtMs: dateSub(
				referenceDate,
				29 * DAY_IN_SECONDS
			).getTime(), // 29 days ago.
		} );

		const { container } = render( <AdsConversionIDSettingsNotice />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should not render if the notice has been dismissed', () => {
		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedItems( [
				ADS_CONVERSION_ID_NOTICE_DISMISSED_ITEM_KEY,
			] );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			adsConversionIDMigratedAtMs: dateSub(
				referenceDate,
				7 * DAY_IN_SECONDS
			).getTime(), // 7 days ago.
		} );

		const { container } = render( <AdsConversionIDSettingsNotice />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should render the notice', () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			adsConversionIDMigratedAtMs: dateSub(
				referenceDate,
				7 * DAY_IN_SECONDS
			).getTime(), // 7 days ago.
		} );

		const { container } = render( <AdsConversionIDSettingsNotice />, {
			registry,
		} );

		expect( container ).toHaveTextContent(
			'Ads Conversion ID has been moved to Ads settings'
		);

		expect( container ).toMatchSnapshot();
	} );
} );
