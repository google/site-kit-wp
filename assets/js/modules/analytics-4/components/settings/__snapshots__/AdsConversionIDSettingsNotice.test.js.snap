// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdsConversionIDSettingsNotice should render the notice 1`] = `
<div>
  <div
    class="googlesitekit-settings-analytics-ads-conversion-id-notice googlesitekit-settings-notice googlesitekit-settings-notice--info googlesitekit-settings-notice--single-row"
  >
    <div
      class="googlesitekit-settings-notice__icon"
    >
      <svg />
    </div>
    <div
      class="googlesitekit-settings-notice__body"
    >
      <div
        class="googlesitekit-settings-notice__text"
      >
        Ads Conversion ID has been moved to
        <a
          class="googlesitekit-cta-link"
          href="undefined#/connected-services/ads"
        >
          <span
            class="googlesitekit-cta-link__contents"
          >
            Ads settings
          </span>
        </a>
      </div>
    </div>
    <div
      class="googlesitekit-settings-notice__button"
    >
      <button
        class="mdc-button mdc-button--tertiary"
        target="_self"
      >
        <span
          class="mdc-button__label"
        >
          Got it
        </span>
      </button>
    </div>
  </div>
</div>
`;
